File: arkcompiler/ets_frontend/ts2panda/src/statement/returnStatement.ts
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as ts from "typescript";
import { CacheList, getVregisterCache } from "../base/vregisterCache";
import { Compiler, ControlFlowChange } from "../compiler";
import { AsyncFunctionBuilder } from "../function/asyncFunctionBuilder";
import { AsyncGeneratorFunctionBuilder } from "../function/asyncGeneratorFunctionBuilder";
import { Label, VReg } from "../irnodes";
import * as jshelpers from "../jshelpers";
import { checkValidUseSuperBeforeSuper } from "./classStatement";

export function compileReturnStatement(stmt: ts.ReturnStatement, compiler: Compiler) {
    let pandaGen = compiler.getPandaGen();
    let returnValue = pandaGen.getTemp();

    if (isReturnInDerivedCtor(stmt)) {
        compileReturnInDerivedCtor(stmt, returnValue, compiler);
    } else {
        compileNormalReturn(stmt, returnValue, compiler);
    }
    pandaGen.freeTemps(returnValue);
}

function compileReturnInDerivedCtor(stmt: ts.ReturnStatement, returnValue: VReg, compiler: Compiler) {
    let pandaGen = compiler.getPandaGen();
    let expr = stmt.expression;

    let need2CheckSuper = pandaGen.getTemp();

    if (!expr) {
        pandaGen.moveVreg(stmt, need2CheckSuper, getVregisterCache(pandaGen, CacheList.True));
    } else {
        if (ts.isCallExpression(expr) && expr.expression.kind == ts.SyntaxKind.SuperKeyword) {
            compileNormalReturn(stmt, returnValue, compiler);
            pandaGen.freeTemps(need2CheckSuper);
            return;
        }

        if (expr.kind == ts.SyntaxKind.ThisKeyword) {
            pandaGen.moveVreg(stmt, need2CheckSuper, getVregisterCache(pandaGen, CacheList.True));
        } else {
            compiler.compileExpression(expr);
            pandaGen.binary(stmt, ts.SyntaxKind.EqualsEqualsEqualsToken, getVregisterCache(pandaGen, CacheList.undefined));
            pandaGen.storeAccumulator(stmt, need2CheckSuper);
        }
    }

    let compile = new Label();
    let notCompile = new Label();
    pandaGen.loadAccumulator(stmt, need2CheckSuper);
    pandaGen.condition(stmt, ts.SyntaxKind.ExclamationEqualsEqualsToken, getVregisterCache(pandaGen, CacheList.False), compile);

    // load this
    let thisReg = pandaGen.getTemp();
    compiler.getThis(stmt, thisReg);
    pandaGen.loadAccumulator(stmt, thisReg);
    pandaGen.branch(stmt, notCompile);

    // compile return expression
    pandaGen.label(stmt, compile);
    if (expr) {
        compiler.compileExpression(expr);
    } else {
        pandaGen.loadAccumulator(stmt, getVregisterCache(pandaGen, CacheList.undefined));
    }

    pandaGen.label(stmt, notCompile);
    pandaGen.storeAccumulator(stmt, returnValue);

    compiler.compileFinallyBeforeCFC(
        undefined,
        ControlFlowChange.Break,
        undefined
    );

    let returnLabel = new Label();
    let normalLabel = new Label();
    pandaGen.loadAccumulator(stmt, need2CheckSuper);
    pandaGen.condition(stmt, ts.SyntaxKind.ExclamationEqualsEqualsToken, getVregisterCache(pandaGen, CacheList.False), normalLabel);
    // check if super has been called
    checkValidUseSuperBeforeSuper(compiler, stmt);
    pandaGen.branch(stmt, returnLabel);

    pandaGen.label(stmt, normalLabel);
    // load returnValue to acc
    pandaGen.loadAccumulator(stmt, returnValue);

    pandaGen.label(stmt, returnLabel);
    pandaGen.return(stmt);

    pandaGen.freeTemps(need2CheckSuper, thisReg);
}

function compileNormalReturn(stmt: ts.ReturnStatement, returnValue: VReg, compiler: Compiler) {
    let expr = stmt.expression;
    let pandaGen = compiler.getPandaGen();
    let empty : boolean = false;

    if (expr) {
        compiler.compileExpression(expr);
    } else {
        empty = true;
        pandaGen.loadAccumulator(stmt, getVregisterCache(pandaGen, CacheList.undefined));
    }
    pandaGen.storeAccumulator(stmt, returnValue);

    compiler.compileFinallyBeforeCFC(
        undefined,
        ControlFlowChange.Break,
        undefined
    );

    pandaGen.loadAccumulator(stmt, returnValue);
    compiler.getFuncBuilder().explicitReturn(stmt, empty);
}

function isReturnInDerivedCtor(stmt: ts.ReturnStatement) {
    let funcNode = jshelpers.getContainingFunctionDeclaration(stmt);
    if (!funcNode || !ts.isConstructorDeclaration(funcNode)) {
        return false;
    }

    if (funcNode && funcNode.parent) {
        let classNode = <ts.ClassLikeDeclaration>funcNode.parent;
        if (classNode.heritageClauses) {
            return true;
        }
    }

    return false;
}

File: arkcompiler/ets_frontend/es2panda/ir/statements/returnStatement.cpp
/**
 * Copyright (c) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "returnStatement.h"

#include <ir/base/methodDefinition.h>
#include <ir/base/scriptFunction.h>
#include <compiler/core/pandagen.h>
#include <typescript/checker.h>

#include <ir/astDump.h>
#include <ir/expression.h>
#include <ir/typeNode.h>

namespace panda::es2panda::ir {

void ReturnStatement::Iterate(const NodeTraverser &cb) const
{
    if (argument_) {
        cb(argument_);
    }
}

void ReturnStatement::Dump(ir::AstDumper *dumper) const
{
    dumper->Add({{"type", "ReturnStatement"}, {"argument", AstDumper::Nullable(argument_)}});
}

void ReturnStatement::Compile(compiler::PandaGen *pg) const
{
    if (argument_) {
        argument_->Compile(pg);
    } else {
        pg->LoadConst(this, compiler::Constant::JS_UNDEFINED);
    }

    if (pg->CheckControlFlowChange()) {
        compiler::RegScope rs(pg);
        compiler::VReg res = pg->AllocReg();

        pg->StoreAccumulator(this, res);
        pg->ControlFlowChangeReturn();
        pg->LoadAccumulator(this, res);
    }

    if (argument_) {
        pg->ValidateClassDirectReturn(this);
        pg->ExplicitReturn(this);
    } else {
        pg->ImplicitReturn(this);
    }
}

checker::Type *ReturnStatement::Check(checker::Checker *checker) const
{
    const ir::AstNode *ancestor = checker::Checker::FindAncestorGivenByType(this, ir::AstNodeType::SCRIPT_FUNCTION);
    ASSERT(ancestor && ancestor->IsScriptFunction());
    const ir::ScriptFunction *containingFunc = ancestor->AsScriptFunction();

    if (containingFunc->Parent()->Parent()->IsMethodDefinition()) {
        const ir::MethodDefinition *containingClassMethod = containingFunc->Parent()->Parent()->AsMethodDefinition();
        if (containingClassMethod->Kind() == ir::MethodDefinitionKind::SET) {
            checker->ThrowTypeError("Setters cannot return a value", Start());
        }
    }

    if (containingFunc->ReturnTypeAnnotation()) {
        checker::Type *returnType = checker->GlobalUndefinedType();
        checker::Type *funcReturnType = containingFunc->ReturnTypeAnnotation()->AsTypeNode()->GetType(checker);

        if (argument_) {
            checker->ElaborateElementwise(funcReturnType, argument_, Start());
            returnType = checker->CheckTypeCached(argument_);
        }

        checker->IsTypeAssignableTo(returnType, funcReturnType,
                                    {"Type '", returnType, "' is not assignable to type '", funcReturnType, "'."},
                                    Start());
    }

    return nullptr;
}

void ReturnStatement::UpdateSelf(const NodeUpdater &cb, [[maybe_unused]] binder::Binder *binder)
{
    if (argument_) {
        argument_ = std::get<ir::AstNode *>(cb(argument_))->AsExpression();
    }
}

}  // namespace panda::es2panda::ir
